### DISCUSSION TITLE
Using enums under tstype

### DISCUSSION THREAD

**Discussion by jonamsalem** [gzuidhof/tygo#51] [Q&A] [answered]
Hey! Is there any way to use my enum under the tstype label in a struct. For example:

say i have an enum :

```
type Status string
const (
    Pending  Status = "Pending"
    Active   Status = "Active"
    Inactive Status = "Inactive"
    Unknown  Status = "Unknown"
)
```
and under my struct i wanna use 
`Status          Status(or String)                json:"status,omitempty" tstype:"Status" `
	
rather than having to generate the status label as` tstype:" 'Pending' | 'Active' | 'Inactive' | 'Unknown' ";`

**Comment by gzuidhof** [comment, marked as answer]
Hi @jonamsalem 

I don't think this is possible directly, it is hard/impossible to know for the current parser to know all the possible values, at least in its current state (it would have to be a more full compiler).

Here is a posisble workaround, you put
```go
type Status string
```
in a separate file that you ignore in the Tygo config.

Then in the frontmatter, you manually write out the type:
```yaml
frontmatter: |
      type Status = "'Pending' | 'Active' | 'Inactive' | 'Unknown'";
```

Or if you want to be even more strict about it:
```typescript
  type Status = "typeof Pending | typeof Active | typeof Inactive | typeof Unknown";
```
Then you would be able to use `tstype:"Status"`.
Of course it's not perfect - but this may just get the job done :).

If your union consists of 1000 items, this is of course not really workable - but maybe it's doable to *type* it manually.

**Reply by jonamsalem** [reply]
I see. Thank you so much!

**Reply by slugger7** [reply]
Although this is an option I hate having to type out the values of my "enums" manually.

Tried to use the `//tygo:emit` but I cant add the values in programatically.
I thought of having a list of all of the "enums" and then somehow getting them all into a type with the pipe separator.

```go
type SomeEnumType = string
const (
  EnumOne SomeEnumType = "EnumOne"
  EnumTwo SomeEnumType = "EnumTwo"
)

var SomeEnumTypeAllValues = []SomeEnumType{
  EnumOne,
  EnumTwo,
}
```

```typescript
export type SomeEnumTypeAllValues = "EnumOne" | "EnumTwo"
```

Would it maybe be possible to add a tag above the list of all the enums in order to get them into typescript how we expect them?

Something else I thought of is to maybe just have a program that outputs these types for me then it will just have to be imported into the rest of the types as frontmatter. Is there then a way to maybe have a pre run task or something like that for instance?

**Reply by slugger7** [reply]
My current solution to this problem is this little program alongside my actual program.
Hope it helps.

https://github.com/slugger7/exorcist-server/blob/main/cmd/enum-export/main.go
